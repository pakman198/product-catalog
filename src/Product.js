import './Product.scss';
import React, {  Component } from 'react';

import CatalogContext from './CatalogContext';

import { capitalize, localizeDate } from './utils';

class Product extends Component {

  renderTags() {
    const { tags } = this.props.data;

    return tags.map(({ title }) => {
      return (
        <button 
          key={title} 
          className="ui teal basic button"
          style={{ marginBottom: '.25em' }}
        >
          { title }
        </button>
      );
    });
  }

  renderProduct(context) {
    const { urls, description, alt_description, created_at, id } = this.props.data;
    
    return (
      <div className="ui card">
        <div className="image" onClick={() => context.fetchPhoto(id)}>
          <img src={urls.regular} alt={alt_description} />
        </div>
        <div className="content">
          <div className="header">{ capitalize(alt_description) }</div>
          <div className="meta">
            <span className="date">{ localizeDate(created_at) }</span>
          </div>
          <div className="description">
            { capitalize(description) }
          </div>
        </div>
        <div className="extra content">
            { this.renderTags() }
        </div>
      </div>
    );

  }
  
  render() {
    return (
      <CatalogContext.Consumer>
        { context => (
            <div>
              { this.renderProduct(context) }
            </div>
          )
        }
      </CatalogContext.Consumer>
    );
  }
}

export default Product;