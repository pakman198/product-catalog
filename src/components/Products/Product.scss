.card {
  .image:hover {
    cursor: pointer;
  }
}