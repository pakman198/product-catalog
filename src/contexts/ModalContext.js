import React, { useState } from 'react';

export const ModalContext = React.createContext();

export const ModalProvider = (props) => {
  const [ isModalVisible, setIsModalVisible ] = useState(false);
  const [ currentItem, setCurrentItem ] = useState({});
  

  const toggleModal = (item) => {
    const selectedItem = item ? item : null;

    setIsModalVisible(!isModalVisible);
    setCurrentItem(selectedItem);
  }

  const contextValue = {
    currentItem,
    isModalVisible,
    toggleModal
  }
    
  return (
    <ModalContext.Provider value={contextValue}>
      { props.children }
    </ModalContext.Provider>
  );
}

