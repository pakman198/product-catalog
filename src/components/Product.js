import './Product.scss';
import React, { useContext } from 'react';

import ModalContext from '../ModalContext';

const currency = (num) => {
  return new Intl.NumberFormat('en-US', {
    style: 'currency',
    currency: 'USD',
    minimumFractionDigits: 2
  }).format(num)
}

const Product = (props) => {
  const context = useContext(ModalContext);
  // console.log({context})
  const { data, data: { name, hero }} = props;

  const price = data.price || data.priceRange;
  const selling = typeof price.selling === 'number' ? price.selling : price.selling.low;


  return (
    <div className="ui centered card">
      <div className="image" onClick={() => context.toggleModal(data)} role="button">
        <img src={hero.href} alt={name} />
      </div>
      <div className="content">
        <div className="header">{name}</div>
      </div>
      <div className="extra content">
          {currency(selling)}
      </div>
    </div>
  );
}

export default Product;