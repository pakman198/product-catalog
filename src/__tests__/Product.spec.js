import React from 'react';
import { 
  render, 
  waitForElement, 
  fireEvent,
} from '@testing-library/react'
import '@testing-library/jest-dom';

import Product from '../components/Product';
import ModalContext from '../ModalContext';
import data from '../__mocks__/apiResponse.json';

describe('Product', () => {
  const product1 = data[0];
  const product2 = data[1];
  const context = {
    toggleModal: jest.fn()
  }

  test('Rendered Properly', () => {
    const { asFragment } = render(
      <ModalContext.Provider value={context}>
        <Product data={product1}/>
      </ModalContext.Provider>
    )

    expect(asFragment()).toMatchSnapshot();
  });

  test('Price comes as priceRange property form the data prop', async () => {
    const { getByText } = render(
      <ModalContext.Provider value={context}>
        <Product data={product2}/>
      </ModalContext.Provider>
    )

    const price  = await waitForElement(() => getByText(/559.20/i));

    expect(price).toBeInTheDOM;
  });

  test('Triggers conext property method when image clicked', async () => {
    const { getByRole } = render(
      <ModalContext.Provider value={context}>
        <Product data={product1}/>
      </ModalContext.Provider>
    )
    const image = await waitForElement(() => getByRole('button'));

    fireEvent.click(image);

    expect(context.toggleModal).toHaveBeenCalledTimes(1);
  });

});