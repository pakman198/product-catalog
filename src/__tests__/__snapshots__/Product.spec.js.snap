// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Rendered Properly 1`] = `
<DocumentFragment>
  <div
    class="ui centered card"
  >
    <div
      class="image"
      role="button"
    >
      <img
        alt="Tahoe Solid Wood Dining Table - Natural Oak"
        src="https://www.westelm.com/weimgs/rk/images/wcm/products/201938/0025/tahoe-solid-wood-dining-table-natural-oak-m.jpg"
      />
    </div>
    <div
      class="content"
    >
      <div
        class="header"
      >
        Tahoe Solid Wood Dining Table - Natural Oak
      </div>
    </div>
    <div
      class="extra content"
    >
      $1,359.20
    </div>
  </div>
</DocumentFragment>
`;
